### ISSUE TITLE
grafana_dashboard fails with Grafana v5.0

### ISSUE THREAD

**Issue by rrey** [ansible/ansible#39663]
##### ISSUE TYPE
 - Bug Report

##### COMPONENT NAME
grafana_dashboard

##### ANSIBLE VERSION
```
ansible 2.5.0
  config file = /Users/remirey/dev/grtgas/avignon/ansible.cfg
  configured module search path = [u'/Users/remirey/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /Users/remirey/.virtualenvs/avignon/lib/python2.7/site-packages/ansible
  executable location = /Users/remirey/.virtualenvs/avignon/bin/ansible
  python version = 2.7.13 (default, Jul 18 2017, 09:17:00) [GCC 4.2.1 Compatible Apple LLVM 8.1.0 (clang-802.0.42)]
```

##### CONFIGURATION
```
DEFAULT_GATHERING(/Users/remirey/dev/grtgas/avignon/ansible.cfg) = explicit
DEFAULT_HASH_BEHAVIOUR(/Users/remirey/dev/grtgas/avignon/ansible.cfg) = replace
DEFAULT_MANAGED_STR(/Users/remirey/dev/grtgas/avignon/ansible.cfg) = Please do not change this file directly since it is managed by Ansible and will be overwritten
DEFAULT_VAULT_PASSWORD_FILE(/Users/remirey/dev/grtgas/avignon/ansible.cfg) = /Users/remirey/dev/grtgas/avignon/ansible_vault_password
HOST_KEY_CHECKING(/Users/remirey/dev/grtgas/avignon/ansible.cfg) = False
RETRY_FILES_ENABLED(/Users/remirey/dev/grtgas/avignon/ansible.cfg) = False
```

##### OS / ENVIRONMENT
N/A

##### SUMMARY

When executing a dashboard creation to test idempotency, the second execution fails with the following error:

```
The full traceback is:
  File "/var/folders/nx/br96pdj13gd30xmz81_fpsw00000gn/T/ansible_NgOhbA/ansible_module_grafana_dashboard.py", line 307, in main
    result = grafana_create_dashboard(module, module.params)
  File "/var/folders/nx/br96pdj13gd30xmz81_fpsw00000gn/T/ansible_NgOhbA/ansible_module_grafana_dashboard.py", line 215, in grafana_create_dashboard
    raise GrafanaAPIException('Unable to create the new dashboard %s : %s - %s.' % (slug, info['status'], info))

failed: [localhost] (item={u'path': u'aws_ecs.json', u'slug': u'aws_ecs'}) => {
    "changed": false,
    "invocation": {
        "module_args": {
            "grafana_api_key": null,
            "grafana_password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
            "grafana_url": "<some_url>",
            "grafana_user": "<some_user>",
            "message": "Updated by ansible",
            "org_id": 1,
            "overwrite": true,
            "path": "files/grafana/dashboards/aws_ecs.json",
            "slug": "aws_ecs",
            "state": "present",
            "validate_certs": true
        }
    },
    "item": {
        "path": "aws_ecs.json",
        "slug": "aws_ecs"
    },
    "msg": "error : Unable to create the new dashboard aws_ecs : 412 - {'body': '{\"message\":\"The dashboard has been changed by someone else\",\"status\":\"version-mismatch\"}', 'date': 'Thu, 03 May 2018 09:37:05 GMT', 'content-length': '88', 'status': 412, 'url': '<some_url>/api/dashboards/db', Expires=Thu, 10 May 2018 09:37:05 GMT; Path=/', 'content-type': 'application/json', 'connection': 'close', 'msg': 'HTTP Error 412: Precondition Failed'}."
```

##### STEPS TO REPRODUCE

```
    - name: Import the Grafana dashboards
      grafana_dashboard:
        grafana_url: <some_url>
        grafana_user: "{{ grafana_admin_user }}"
        grafana_password: "{{ grafana_admin_password }}"
        state: present
        message: Updated by ansible
        overwrite: yes
        path: "files/grafana/dashboards/{{ item.path }}"
        slug: "{{ item.slug }}"
      with_items:
        - path: aws_ecs.json
          slug: aws_ecs
```

The meta section was added manually in the dashboard json to have the import work in the first place.

##### EXPECTED RESULTS
Idempotency works and second execution marks the task "ok" and not "changed".

##### ACTUAL RESULTS
Exception

```
The full traceback is:
  File "/var/folders/nx/br96pdj13gd30xmz81_fpsw00000gn/T/ansible_NgOhbA/ansible_module_grafana_dashboard.py", line 307, in main
    result = grafana_create_dashboard(module, module.params)
  File "/var/folders/nx/br96pdj13gd30xmz81_fpsw00000gn/T/ansible_NgOhbA/ansible_module_grafana_dashboard.py", line 215, in grafana_create_dashboard
    raise GrafanaAPIException('Unable to create the new dashboard %s : %s - %s.' % (slug, info['status'], info))
```

**Comment by ansibot** [comment]
Files identified in the description:
* [lib/ansible/modules/monitoring/grafana_dashboard.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/monitoring/grafana_dashboard.py)

If these files are inaccurate, please update the `component name` section of the description or use the `!component` bot command.

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
cc @tsalle
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by rrey** [comment]
I dug a bit in the module code, and the module use the slug to retrieve the dashboard:

```
def grafana_dashboard_exists(module, grafana_url, slug, headers):
    dashboard_exists = False
    dashboard = {}
    r, info = fetch_url(module, '%s/api/dashboards/db/%s' % (grafana_url, slug), headers=headers, method='GET')
    if info['status'] == 200:
        dashboard_exists = True
        try:
            dashboard = json.loads(r.read())
        except Exception as e:
            raise GrafanaMalformedJson(e)
    elif info['status'] == 404:
        dashboard_exists = False
    else:
        raise GrafanaAPIException('Unable to get dashboard %s : %s' % (slug, info))

    return dashboard_exists, dashboard
```

In Grafana v5, the "Get dashboard by slug" is deprecated, and I could see that the dashboard json does not contain the slug even if it was present in the imported json used by the ansible module.
It leads to the dashboard retrieval to get a 404, and the dashboard is not found.

It may be related to the final error, but I didn't go further yet.

**Comment by cortexx** [comment]
Hi,

Same problem here and I confirm that in Grafana release number `>= 5.0` the method `Get dashboard by slug` had been deprecated. The method `Get dashboard by uid resource` is recommended instead.

See link for more information : http://docs.grafana.org/http_api/dashboard/#get-dashboard-by-slug

Do you know any workaround to implement to continue to use this useful module ?

##### GRAFANA VERSION
```
5.0.4-1
```
##### ANSIBLE VERSION
```
ansible 2.5.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/${USER}/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/site-packages/ansible-2.5.0-py2.7.egg/ansible
  executable location = /usr/bin/ansible
  python version = 2.7.5 (default, Nov  6 2016, 00:28:07) [GCC 4.8.5 20150623 (Red Hat 4.8.5-11)]
```
##### COMPONENT NAME
grafana_dashboard
##### OS / ENVIRONMENT
Running Ansible from : `CentOS Linux release 7.3.1611 (Core)`
OS managing by Ansible : `CentOS Linux release 7.3.1611 (Core)`

Thanks !

**Comment by rrey** [comment]
I guess we need a patch for this. I didn't find time to start working on it yet.

cc @seuf by the way.

**Comment by packetninja** [comment]
I ran into this as well. Unfortunately I lack the knowledge to fix this myself, I'll work around it statically. Thanks for the work @rrey

**Comment by seuf** [comment]
Ok, I will take a look this week..

**Comment by rrey** [comment]
@seuf I took some time to check how a fix could be made. I guess that we should have both methods available (slug and uuid) to allow the module to work with early versions of Grafana and more recent ones.

I saw that there are some code duplication in the module (for the header management for instance) and that authentication is not taking advantage of the mechanisms available in fetch_url helper.

Maybe it is an occasion to do a bit of refactoring ? 
I can try to push something this weekend if you are interested ?
